### PR TITLE
nixos/tests/test-driver: add shell_interact

### PR DESCRIPTION
**Pull request by Synthetica9**
###### Motivation for this change

Currently, there was no easy way to drop into a guest shell from the test driver. This PR adds the method `machine.shell_interact()`, which drops you into the "main" guest shell. 
A few notes:
* This isn't super polished, you get a very bare-bones shell without anything pretty like tab completion.
* Killing the interactive session with `ctrl-d` or `ctrl-c` also ends the guest session.

A better implementation might be to set up a real ssh session, which would aleviate the above two issues. However, that would also be a lot more work than these 4 lines of code.

However, I think this is still worth adding since this is not very discoverable (even though it is pretty much built-in to python).

Open questions:

- [x] Should this be backported to 20.05?
- [ ] Is this important enough for a release note?

cc @domenkozar

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lheckemann** [comment]
>Currently, there was no easy way to drop into a guest shell from the test driver.

I used vncviewer to connect to the VNC display exposed by qemu.

Regardless, this is nice! Things like the clipboard are nice to have :)

**Review by andir — COMMENTED** [review]
Please also document the function in `manual/development/writing-nixos-tests.xml` just to make sure it is more discoverable :-)

I've been thinking of doing some changes to the test runner for a while now. My main goal was, just as yours, to obtain a shell in the VMs without running them in graphical mode. The way I was thinking about implementing this was as Tmux session that would have a one "pane"(is that the right term?) for each VMs serial console. This would require adding a 2nd serial TTY to the VMs to not interfere with the test runner but that should not be harmful.

We could even have the default view be a split between machines spamming their dmesg/syslog and a clean Python shell (instead of the interleaved output we have now).

File: nixos/doc/manual/development/writing-nixos-tests.xml [hunk 1]
```diff
@@ -436,6 +436,17 @@ machine.systemctl("list-jobs --no-pager", "any-user") # spawns a shell for `any-
...
     </listitem>
    </varlistentry>
+   <varlistentry>
+    <term>
+     <methodname>shell_interact</methodname>
+    </term>
+    <listitem>
+     <para>
+      Allows you to directly interact with the guest shell.
+      This should only be used during test development, not in production tests.
```
Comment (andir) [inline nixos/doc/manual/development/writing-nixos-tests.xml:446]: Perhaps mention:
```
Killing the interactive session with ctrl-d or ctrl-c also ends the guest session.
```

**Comment by domenkozar** [comment]
Fixes https://github.com/NixOS/nixpkgs/issues/125359 and I agree with @andir that adding just a sentence or two in narrative documentation would make a lot of beginners happy.

**Comment by tfc** [comment]
does anything speak against writing a function which accepts a machine as parameter and then runs this, and which is only defined in shell context and not nix-build context?

My central fear is that the `Machine` class grows more and more complex and due to Python's nature of everything being public automatically etc. everything we add becomes the defacto-Machine-interface no matter if designed for this or not.

**Comment by domenkozar** [comment]
> does anything speak against writing a standalone function which accepts a machine as parameter and then runs this, and which is only defined in shell context and not nix-build context?
> 
> My central fear is that the `Machine` class grows more and more complex and due to Python's nature of everything being public automatically etc. everything we add becomes the defacto-Machine-interface no matter if designed for this or not.

I share these concerns, but I don't think we should block this work as improving debugging is crucial and exposing this as public makes sense (to me).

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 125372` at 28714427 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.trivial</li>
  </ul>
</details>

**Comment by blaggacao** [comment]
> The way I was thinking about implementing this was as Tmux session that would have a one "pane"(is that the right term?) for each VMs serial console.

> We could even have the default view be a split between machines spamming their dmesg/syslog and a clean Python shell (instead of the interleaved output we have now).

I'm currently working on some prototype with `tmux`. My main goal is split logs per unit for display only, but interleaving with the idea of this PR might be in order.

**Comment by tfc** [comment]
> I share these concerns, but I don't think we should block this work as improving debugging is crucial and exposing this as public makes sense (to me).

yes, debugging is crucial and i wouldn't make a big deal about finding the perfect implementation but in my naive mind it should be easy to define

```python
def interact(m: Machine) -> void:
  m.connect()
  telnet = telnetlib.Telnet()
  telnet.sock = m.shell  # type: ignore        
  telnet.interact()
```

on the path that results in the python repl and mention the function in the docs? if this is not possible for good reasons i turn in, but if it is possible, then it also seems cleaner to me.

**Comment by tfc** [comment]
> > The way I was thinking about implementing this was as Tmux session that would have a one "pane"(is that the right term?) for each VMs serial console.
> 
> > We could even have the default view be a split between machines spamming their dmesg/syslog and a clean Python shell (instead of the interleaved output we have now).
> 
> I'm currently working on some prototype with `tmux`. My main goal is split logs per systemd-unit for display only, but interleaving with the idea of this PR might be in order.

there were earlier discussions where we converged towards a nice design but it stalled on the implementation due to lack of time (not sure). that discussion was between me, @flokli and @nlewo  (it was his initial idea).
maybe we should sync on that, too.

**Comment by blaggacao** [comment]
@tfc Cool, please join them and yourself in the matrix room, too. :wink:

**Comment by tfc** [comment]
> @tfc Cool, please join them and yourself in the matrix room, above, too. 

i am there already. unfortunately over the day i am often stuck in meetings and can only reply in the evenings or the day after, so such discussions are better for me to have in person in a videocall or some more asynchronous medium than chat.

**Comment by blaggacao** [comment]
Recording outcomes:

We had a little bouncing of ideas in the chat, @nlewo pointed to prior art that they had done which was improved by @tfc [here](https://github.com/tfc/nixpkgs/tree/tmux-test-driver/nixos/lib/test-driver).

As a consequence of this brainstorming, I started cleanup work in #125418, which we hope should us enable to implement a very nice implementation of host access in a relatively short timespan. 

@nlewo shared an asciinema showcase of what we are aiming for, here: https://asciinema.org/a/8a1wfDnzWHWfff2AHqSmGeRV0

Amazing!

**Comment by domenkozar** [comment]
I'd still like to merge this as tmux is not exactly beginner-friendly (not everyone knows how to use tmux). 

@Synthetica9 do you plan to implement the prompt or do we merge it as-is?

**Comment by Synthetica9** [comment]
> I'd still like to merge this as tmux is not exactly beginner-friendly (not everyone knows how to use tmux).
> 
> @Synthetica9 do you plan to implement the prompt or do we merge it as-is?

I think a merge as-is is fine, if we want proper shell support we should probably do it on a different backend (maybe we can plug into the tmux backend even if not using tmux?)

File: nixos/lib/test-driver/test-driver.py [hunk 2]
```diff
@@ -455,6 +456,15 @@ def execute(self, command: str) -> Tuple[int, str]:
                 return (status_code, output)
             output += chunk
 
+    def shell_interact(self) -> None:
+        """Allows you to interact with the guest shell
+
+        Should only be used during test development, not in the production test."""
+        self.connect()
```
Comment (domenkozar) [inline nixos/lib/test-driver/test-driver.py:463]:
```suggestion
        self.connect()
        print("Welcome to " + self.name + ":")
```
```

**Comment by domenkozar** [comment]
> > I share these concerns, but I don't think we should block this work as improving debugging is crucial and exposing this as public makes sense (to me).
> 
> yes, debugging is crucial and i wouldn't make a big deal about finding the perfect implementation but in my naive mind it should be easy to define
> 
> ```python
> def interact(m: Machine) -> void:
>   m.connect()
>   telnet = telnetlib.Telnet()
>   telnet.sock = m.shell  # type: ignore        
>   telnet.interact()
> ```
> 
> on the path that results in the python repl and mention the function in the docs? if this is not possible for good reasons i turn in, but if it is possible, then it also seems cleaner to me.

I prefer functional style as well, but for sake of consistency, it's better that either whole Machine API uses OOP or FP style. A good place to change that is in https://github.com/NixOS/nixpkgs/pull/125418

**Comment by blaggacao** [comment]
> maybe we can plug into the tmux backend even if not using tmux?

Excellent idea! I plan to use `byobu` which has an `F1` cheatsheet and a complete `F9` (irrc) help menu.

We could easily advertise the `F1` cheat sheet as the MOTD.

**Comment by Mic92** [comment]
I have a follow up PR: https://github.com/NixOS/nixpkgs/pull/127763
socat's readline is nicer than termlib, which cannot handle systemctl output due to ansi encoding.

**Pull request merged**